### ISSUE TITLE
`UnionType` instance not passable as generic parameter

### ISSUE THREAD

**Issue by tpgillam** [microsoft/pyright#8534]
We've run into this issue whilst trying to upgrade from pyright 1.1.369 to 1.1.373 internally:

```python
from types import UnionType

T1: UnionType = int | str
list[T1]  # Happy

T2: UnionType = T1 | set
list[T2]  # Happy in pyright 1.1.369, unhappy in pyright 1.1.373
```

The error given on the final line in 1.1.373 is:
```
Diagnostics:
1. Expected class but received "UnionType" [reportGeneralTypeIssues]
```

(this is a somewhat contrived minimal example, but I think captures our internal use sufficiently).

We'd expect this to not be an error, since it runs correctly.

**Comment by erictraut** [comment]
Your code sample is relying on behaviors that are not well specified in the typing spec. The fact that it "runs correctly" isn't a good argument because code will often run without generating a runtime exception but still result in type checker errors. The question is whether a subscript expression for a generic class should be required to use "type expression" evaluation rules when outside of a type annotation.

For comparison, mypy rejects both `list[T1]` and `list[T2]`. This is more consistent than pyright's current behavior because it accepts the first but rejects the second.

There's an argument to be made that a type checker should not enforce type expression rules in this case because this is a value expression — i.e. it's not used in a location within the grammar reserved for [type expressions](https://typing.readthedocs.io/en/latest/spec/annotations.html#type-and-annotation-expressions). This is the principle I generally follow in pyright. The behavior you're seeing here is inconsistent with this principle, so I think a change is warranted. This will be included in the next release. With this change, you won't see an error for either `list[T1]` or `list[T2]`.

BTW, draft [PEP 747](https://discuss.python.org/t/pep-747-typeexpr-type-hint-for-a-type-expression/55984/40) is attempting to nail down some of these rules so this is no longer a gray area in the typing spec. This draft still has a ways to go before it's complete, but it's likely that `T1: UnionType = int | str` will become a type error under the rules that it defines. `UnionType` is an internal implementation detail of the type system (similar to `GenericAlias` or `_AnnotatedAlias`). These are not part of the static type system, so they shouldn't generally be used in type expressions. This is just a heads up that your code may need to be modified in the future once the typing spec clarifies this behavior.

**Comment by tpgillam** [comment]
> The question is whether a subscript expression for a generic class should be required to use "type expression" evaluation rules when outside of a type annotation.

Yes exactly; apologies my example didn't make this sufficiently clear. We're currently relying on the ability to construct a `GenericAlias` instance at runtime. (Our application for this is to specify types in an internal serialisation library). We don't actually have code that specifically annotates a variable as `UnionType`, but rather in some of our tests we construct lists of type union types, and then when we iterate over these lists pyright infers the elements to be of type `UnionType`.

Thank you for explaining all the context around this!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
